core/src/core/pmmr/segment.rs

```diff
@@ -0,0 +1,519 @@
+	hashes: Vec<Hash>,
+}
+
+impl SegmentProof {
+	fn empty() -> Self {
+		Self { hashes: Vec::new() }
+	}
+
+	fn generate<U, B>(
+		pmmr: &ReadonlyPMMR<'_, U, B>,
+		last_pos: u64,
+		segment_first_pos: u64,
+		segment_last_pos: u64,
+	) -> Result<Self, SegmentError>
+	where
+		U: PMMRable,
+		B: Backend<U>,
+	{
+		let family_branch = pmmr::family_branch(segment_last_pos, last_pos);
+
+		// 1. siblings along the path from the subtree root to the peak
+		let hashes: Result<Vec<_>, _> = family_branch
+			.iter()
+			.map(|&(_, s)| pmmr.get_hash(s).ok_or_else(|| SegmentError::MissingHash(s)))
+			.collect();
+		let mut proof = Self { hashes: hashes? };
+
+		// 2. bagged peaks to the right
+		let peak_pos = family_branch
+			.last()
+			.map(|&(p, _)| p)
+			.unwrap_or(segment_last_pos);
+		if let Some(h) = pmmr.bag_the_rhs(peak_pos) {
+			proof.hashes.push(h);
+		}
+
+		// 3. peaks to the left
+		let peaks: Result<Vec<_>, _> = pmmr::peaks(last_pos)
+			.into_iter()
+			.filter(|&x| x < segment_first_pos)
+			.rev()
+			.map(|p| pmmr.get_hash(p).ok_or_else(|| SegmentError::MissingHash(p)))
+			.collect();
+		proof.hashes.extend(peaks?);
+
+		Ok(proof)
+	}
+
+	/// Reconstruct PMMR root using this proof
+	pub fn reconstruct_root(
+		&self,
+		last_pos: u64,
+		segment_first_pos: u64,
+		segment_last_pos: u64,
+		segment_root: Hash,
+	) -> Result<Hash, SegmentError> {
+		let mut iter = self.hashes.iter();
+		let family_branch = pmmr::family_branch(segment_last_pos, last_pos);
+
+		// 1. siblings along the path from the subtree root to the peak
+		let mut root = segment_root;
+		for &(p, s) in &family_branch {
+			let sibling_hash = iter.next().ok_or_else(|| SegmentError::MissingHash(s))?;
+			root = if pmmr::is_left_sibling(s) {
+				(sibling_hash, root).hash_with_index(p - 1)
+			} else {
+				(root, sibling_hash).hash_with_index(p - 1)
+			};
+		}
+
+		// 2. bagged peaks to the right
+		let peak_pos = family_branch
+			.last()
+			.map(|&(p, _)| p)
+			.unwrap_or(segment_last_pos);
+
+		let rhs = pmmr::peaks(last_pos)
+			.into_iter()
+			.filter(|&x| x > peak_pos)
+			.next();
+
+		if let Some(pos) = rhs {
+			root = (
+				root,
+				iter.next().ok_or_else(|| SegmentError::MissingHash(pos))?,
+			)
+				.hash_with_index(last_pos)
+		}
+
+		// 3. peaks to the left
+		let peaks = pmmr::peaks(last_pos)
+			.into_iter()
+			.filter(|&x| x < segment_first_pos)
+			.rev();
+		for pos in peaks {
+			root = (
+				iter.next().ok_or_else(|| SegmentError::MissingHash(pos))?,
+				root,
+			)
+				.hash_with_index(last_pos);
+		}
+
+		Ok(root)
+	}
+
+	/// Check validity of the proof by equating the reconstructed root with the actual root
+	pub fn validate(
+		&self,
+		last_pos: u64,
+		mmr_root: Hash,
+		segment_first_pos: u64,
+		segment_last_pos: u64,
+		segment_root: Hash,
+	) -> Result<(), SegmentError> {
+		let root =
+			self.reconstruct_root(last_pos, segment_first_pos, segment_last_pos, segment_root)?;
+		if root == mmr_root {
```

Comment:
We commit to the `output_root`, `range_proof_root` and `kernel_root` in the block header.
We can pass the `range_proof_root` and `kernel_root` into `validate()` directly and confirm the root matches.

The `output_root` needs some thinking about as this is a merged root of both the output PMMR and the bitmap accumulator (they are hashed together).
So we cannot validate one is correct without providing the other.

If we receive a segment of the bitmap accumulator then I think we need to be provided with the corresponding output root to allow validation.
And conversely if we receive a segment of the output PMMR then we need the corresponding bitmap root.

Its not clear to me if this should be provided as part of the segment proof itself or provided alongside it.
There is some context specific logic that needs to occur there to know what kind of segment this is and what additional (if any) root hashes must be provided.

Reply:
> The output_root needs some thinking about as this is a merged root of both the output PMMR and the bitmap accumulator (they are hashed together).
So we cannot validate one is correct without providing the other.

Yes, good point. So maybe it would make sense to introduce an additional p2p message to request the roots of the output PMMR and the bitmap accumulator. We can validate them by hashing them together to see if they reproduce `output_root` in the header. Then during the validation of segments we can use the roots directly. This sounds like the easiest solution to me, rather than having to alter the structure of segments for this specific situation.

Reply:
I'd personally prefer if these segment response msgs were self-contained and included this _other_ root. So not in the segment itself, but still included in the response msg.
Adding a whole new p2p msg for this seems overly complicated.
That said - I think either solution would work.

Reply:
Ok, I think that makes sense. The PR [here](https://github.com/mimblewimble/grin/pull/3470) defines the actual p2p messages (to be rebased after this one is merged). I will update it to add an extra field for the `OutputBitmapSegment` and `OutputSegment` messages.